src/modules/fancyzones/lib/JsonHelpers.cpp

```diff
@@ -1101,222 +406,211 @@ namespace JSONHelpers
+        {
+            TDeviceInfoMap deviceInfoMap{};
+            auto devices = fancyZonesDataJSON.GetNamedArray(L"devices");
+
+            for (uint32_t i = 0; i < devices.Size(); ++i)
+            {
+                if (auto device = DeviceInfoJSON::DeviceInfoJSON::FromJson(devices.GetObjectAt(i)); device.has_value())
+                {
+                    deviceInfoMap[device->deviceId] = std::move(device->data);
+                }
+            }
+
+            return std::move(deviceInfoMap);
+        }
+        catch (const winrt::hresult_error&)
         {
-            cellRow.resize(m_columns, 0);
+            return {};
         }
     }
 
-    json::JsonObject GridLayoutInfo::ToJson(const GridLayoutInfo& gridInfo)
+    json::JsonArray SerializeDeviceInfos(const TDeviceInfoMap& deviceInfoMap)
     {
-        json::JsonObject infoJson;
-        infoJson.SetNamedValue(L"rows", json::value(gridInfo.m_rows));
-        infoJson.SetNamedValue(L"columns", json::value(gridInfo.m_columns));
-        infoJson.SetNamedValue(L"rows-percentage", NumVecToJsonArray(gridInfo.m_rowsPercents));
-        infoJson.SetNamedValue(L"columns-percentage", NumVecToJsonArray(gridInfo.m_columnsPercents));
+        json::JsonArray DeviceInfosJSON{};
 
-        json::JsonArray cellChildMapJson;
-        for (int i = 0; i < gridInfo.m_cellChildMap.size(); ++i)
+        for (const auto& [deviceID, deviceData] : deviceInfoMap)
         {
-            cellChildMapJson.Append(NumVecToJsonArray(gridInfo.m_cellChildMap[i]));
+            if (deviceData.activeZoneSet.type != FancyZonesDataTypes::ZoneSetLayoutType::Blank)
+            {
+                DeviceInfosJSON.Append(DeviceInfoJSON::DeviceInfoJSON::ToJson(DeviceInfoJSON{ deviceID, deviceData }));
+            }
         }
-        infoJson.SetNamedValue(L"cell-child-map", cellChildMapJson);
 
-        return infoJson;
+        return DeviceInfosJSON;
     }
 
-    std::optional<GridLayoutInfo> GridLayoutInfo::FromJson(const json::JsonObject& infoJson)
+    TCustomZoneSetsMap ParseCustomZoneSets(const json::JsonObject& fancyZonesDataJSON)
     {
         try
         {
-            GridLayoutInfo info(GridLayoutInfo::Minimal{});
-
-            info.m_rows = static_cast<int>(infoJson.GetNamedNumber(L"rows"));
-            info.m_columns = static_cast<int>(infoJson.GetNamedNumber(L"columns"));
-
-            json::JsonArray rowsPercentage = infoJson.GetNamedArray(L"rows-percentage");
-            json::JsonArray columnsPercentage = infoJson.GetNamedArray(L"columns-percentage");
-            json::JsonArray cellChildMap = infoJson.GetNamedArray(L"cell-child-map");
-
-            if (rowsPercentage.Size() != info.m_rows || columnsPercentage.Size() != info.m_columns || cellChildMap.Size() != info.m_rows)
-            {
-                return std::nullopt;
-            }
+            TCustomZoneSetsMap customZoneSetsMap{};
+            auto customZoneSets = fancyZonesDataJSON.GetNamedArray(L"custom-zone-sets");
```

Comment:
Replace all strings literals with const strings and explicitly mark them as non-localizable.